from rattle import Run

RUN = Run(config)

# TODO: ensure db indices exist
RUN.settings["genome_fasta_dict"] = RUN.settings["genome_fasta"].rsplit(".", 1)[0] + ".dict"
RUN.set_default_setting("extract_script", srcdir(path.join("scripts", "vcf2json.py")))
RUN.set_default_setting("csv_script", srcdir(path.join("scripts", "json2csv.py")))
RUN.set_default_setting("plot_script", srcdir(path.join("scripts", "plotVariants.R")))

if RUN.settings.get("gsnap_exe") is not None:
    # To reproduce paper results, relies on a non-conda GSNAP 2014.12.23 with SIMD
    rule align_vars:
        input:
            fq1=RUN.output("{sample}/{sample}-R1.fq.gz"),
            fq2=RUN.output("{sample}/{sample}-R2.fq.gz"),
            index=RUN.settings["genome_gmap_index"],
            exe=RUN.settings["gsnap_exe"],
        output:
            bam=temp(RUN.output("{sample}/snv-indels/{sample}.snv-indel.raw.bam")),
            bai=temp(RUN.output("{sample}/snv-indels/{sample}.snv-indel.raw.bai")),
        params:
            rg_sample="{sample}"
        threads: 8
        conda: srcdir("envs/align_vars.yml")
        shell:
            "{input.exe} --dir `dirname {input.index}` --db `basename {input.index}`"
            " --batch 4 --nthreads {threads}"
            " --novelsplicing 1 --npaths 1 --quiet-if-excessive"
            " --read-group-name={params.rg_sample} --read-group-id={params.rg_sample}"
            " --format sam --gunzip {input.fq1} {input.fq2}"
            " | picard SortSam I=/dev/stdin O={output.bam} SORT_ORDER=coordinate"
            " VALIDATION_STRINGENCY=SILENT CREATE_INDEX=true"

    rule reorder_aln_header:
        input:
            bam=RUN.output("{sample}/snv-indels/{sample}.snv-indel.raw.bam"),
            bai=RUN.output("{sample}/snv-indels/{sample}.snv-indel.raw.bai"),
            ref=RUN.settings["genome_fasta"],
            refd=RUN.settings["genome_fasta_dict"],
        output:
            bam=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bam"),
            bai=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bai"),
        params:
            rg_sample="{sample}"
        threads: 1
        conda: srcdir("envs/align_vars.yml")
        shell:
            "picard -Xmx3G ReorderSam I={input.bam} O={output.bam} R={input.ref}"
            " VALIDATION_STRINGENCY=SILENT CREATE_INDEX=true"
else:
    rule align_vars:
        input:
            fq1=RUN.output("{sample}/{sample}-R1.fq.gz"),
            fq2=RUN.output("{sample}/{sample}-R2.fq.gz"),
            index=RUN.settings["genome_star_index"],
        output:
            bam=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bam"),
            bai=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bai"),
        params:
            rg_sample="{sample}"
        threads: 8
        conda: srcdir("envs/align_vars.yml")
        shell:
            "rm -rf `dirname {output.bam}`/alignment && "
            "mkdir -p `dirname {output.bam}`/alignment && "
            "STAR --outFileNamePrefix `dirname {output.bam}`/alignment/"
            " --genomeDir {input.index} --readFilesIn {input.fq1} {input.fq2}"
            " --readFilesCommand zcat --runThreadN {threads}"
            " --outStd SAM --outSAMunmapped Within --outFilterMultimapNmax 1"
            " --outSAMattrRGline ID:{params.rg_sample} SM:{params.rg_sample}"
            " | picard SortSam I=/dev/stdin O={output.bam} SORT_ORDER=coordinate"
            " VALIDATION_STRINGENCY=SILENT CREATE_INDEX=true"

rule call_vars:
    input:
        bam=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bam"),
        ref=RUN.settings["genome_fasta"],
    output:
        vcf=RUN.output("{sample}/snv-indels/{sample}.raw.vcf.gz")
    threads: 3
    conda: srcdir("envs/call_vars.yml")
    shell:
        "samtools mpileup -f {input.ref} -d 1000000 -s -B {input.bam}"
        " | grep -vP '\\t\\t'"
        " | varscan mpileup2cns --strand-filter 0 --output-vcf 1 --min-var-freq 0.1 --p-value 0.05"
        " | grep -vP '\\t\./\.|\\t0/0'"
        " | bgzip -c > {output.vcf}"

rule aln_stats:
    input:
        bam=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bam"),
        bai=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bai"),
        ref=RUN.settings["genome_fasta"],
        ref_dict=RUN.settings["genome_fasta_dict"],
    output:
        stats=RUN.output("{sample}/snv-indels/{sample}.aln_stats")
    threads: 1
    conda: srcdir("envs/bam_stats.yml")
    shell:
        "picard -Xmx2G CollectAlignmentSummaryMetrics"
        " R={input.ref} I={input.bam} O={output.stats}"

rule insert_stats:
    input:
        bam=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bam"),
        bai=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bai"),
        ref=RUN.settings["genome_fasta"],
        ref_dict=RUN.settings["genome_fasta_dict"],
    output:
        stats=RUN.output("{sample}/snv-indels/{sample}.insert_stats")
        histo=RUN.output("{sample}/snv-indels/{sample}.insert_stats.pdf")
    threads: 1
    conda: srcdir("envs/bam_stats.yml")
    shell:
        "picard -Xmx2G CollectInsertSizeMetrics"
        " R={input.ref} I={input.bam} O={output.stats} H={output.histo}"

rule rna_stats:
    input:
        bam=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bam"),
        bai=RUN.output("{sample}/snv-indels/{sample}.snv-indel.bai"),
        ref=RUN.settings["genome_fasta"],
        ref_dict=RUN.settings["genome_fasta_dict"],
        annot=RUN.settings["annotation_refflat"],
    output:
        stats=RUN.output("{sample}/snv-indels/{sample}.rna_stats")
    threads: 1
    conda: srcdir("envs/bam_stats.yml")
    shell:
        "picard -Xmx4G CollectRnaSeqMetrics"
        " R={input.ref} REF_FLAT={input.annot}"
        " STRAND_SPECIFICITY=NONE ASSUME_SORTED=true"
        " CHART_OUTPUT={output.stats}.pdf I={input.bam} O={output.stats}"

rule annotate_vars:
    input:
        vcf=RUN.output("{sample}/snv-indels/{sample}.raw.vcf.gz"),
        ref_1kg=RUN.settings["vcf_1kg"],
        ref_gonl=RUN.settings["vcf_gonl"],
        cache_vep=RUN.settings["cache_vep"],
        fasta_vep=RUN.settings["fasta_vep"],
        vep_exe=RUN.settings["vep_exe"],
        bgzip_exe=RUN.settings["bgzip_exe"],
    output:
        vcf=RUN.output("{sample}/snv-indels/{sample}.annotated.vcf.gz"),
        stats=RUN.output("{sample}/snv-indels/{sample}.vep.stats"),
    threads: 8
    shell:
        "PATH=/usr/bin:${{PATH}} {input.vep_exe} -i {input.vcf}"
        " --fasta {input.fasta_vep} --dir {input.cache_vep}"
        " --custom {input.ref_1kg},P3,vcf,exact,0,AF,AFR_AF,AMR_AF,EAS_AF,EUR_AF,SAS_AF"
        " --custom {input.ref_gonl},GONL,vcf,exact,0,AF"
        " --fork {threads} --offline --cache_version 77 --assembly GRCh38 --everything --merged"
        " --allele_number --stats_text --vcf --cache --force_overwrite"
        " --stats_file {output.stats} -o STDOUT | {input.bgzip_exe} -c > {output.vcf}"

rule extract_vars:
    input:
        vcf=RUN.output("{sample}/snv-indels/{sample}.annotated.vcf.gz"),
        ref_hotspots=RUN.settings["bed_variant_hotspots"],
        ref_id_mapping=RUN.settings["ref_id_mapping"],
        scr=RUN.settings["extract_script"],
    output:
        json=RUN.output("{sample}/snv-indels/{sample}.variants.json"),
    threads: 1
    conda: srcdir("envs/extract_vars.yml")
    shell:
        "python {input.scr} --hotspots {input.ref_hotspots} --sample-id {wildcards.sample}"
        " {input.ref_id_mapping} {input.vcf} > {output.json}"

rule table_vars_all:
    input:
        json=RUN.output("{sample}/snv-indels/{sample}.variants.json"),
        scr=RUN.settings["csv_script"],
    output:
        csv=RUN.output("{sample}/snv-indels/{sample}.variants_all.csv"),
    threads: 1
    conda: srcdir("envs/json_to_csv.yml")
    shell:
        "python {input.scr} {input.json} > {output.csv}"

rule table_vars_hi:
    input:
        json=RUN.output("{sample}/snv-indels/{sample}.variants.json"),
        scr=RUN.settings["csv_script"],
    output:
        csv=RUN.output("{sample}/snv-indels/{sample}.variants_hi.csv"),
    threads: 1
    conda: srcdir("envs/json_to_csv.yml")
    shell:
        "python {input.scr} --hi {input.json} > {output.csv}"

rule plot_vars:
    input:
        json=RUN.output("{sample}/snv-indels/{sample}.variants.json"),
        scr=RUN.settings["plot_script"],
        ref_id_mapping=RUN.settings["ref_id_mapping"],
        ref_hotspots=RUN.settings["bed_variant_hotspots"],
        ref_amplicons=RUN.settings["ref_amplicons"],
        ref_annot_goi=RUN.settings["ref_annot_goi"],
    params:
        plotd=RUN.output("{sample}/snv-indels/variant_plots"),
    output:
        plots_mark=RUN.output("{sample}/snv-indels/variant_plots/.done"),
    threads: 1
    conda: srcdir("envs/plot_variants.yml")
    shell:
        "(Rscript {input.scr} -j {input.json}"
        " -i {input.ref_id_mapping} -m {input.ref_annot_goi}"
        " -a {input.ref_amplicons} -t {input.ref_hotspots}"
        " -o `dirname {params.plotd}` && touch {params.plotd}/.done)"
        " || rm -rf {params.plotd}"
